pinecone/client.go

```diff
@@ -463,16 +476,75 @@ func toCollection(cm *control.CollectionModel) *Collection {
 	}
 }
 
-func decodeCollection(resBody io.ReadCloser) (*Collection, error) {
+func decodeCollection(resBodyBytes []byte) (*Collection, error) {
 	var collectionModel control.CollectionModel
-	err := json.NewDecoder(resBody).Decode(&collectionModel)
+	err := json.Unmarshal(resBodyBytes, &collectionModel)
 	if err != nil {
 		return nil, fmt.Errorf("failed to decode collection response: %w", err)
 	}
 
 	return toCollection(&collectionModel), nil
 }
 
+func decodeErrorResponse(resBodyBytes []byte) (*control.ErrorResponse, error) {
+	var errorResponse control.ErrorResponse
+	err := json.Unmarshal(resBodyBytes, &errorResponse)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode error response: %w", err)
+	}
+
+	if errorResponse.Status == 0 {
+		return nil, fmt.Errorf("unable to parse ErrorResponse: %v", string(resBodyBytes))
+	}
+
+	return &errorResponse, nil
+}
+
+type errorResponseMap struct {
+	StatusCode int    `json:"status_code"`
+	Body       string `json:"body,omitempty"`
+	ErrorCode  string `json:"error_code,omitempty"`
+	Message    string `json:"message,omitempty"`
+	Details    string `json:"details,omitempty"`
+}
+
+func handleErrorResponseBody(resBodyBytes []byte, statusCode int, errMsgPrefix string) error {
+	var errMap errorResponseMap
+
+	errMap.StatusCode = statusCode
+
+	// try and decode ErrorResponse
+	if json.Valid(resBodyBytes) {
+		errorResponse, err := decodeErrorResponse(resBodyBytes)
+		if err == nil {
+			errMap.Message = errorResponse.Error.Message
+			errMap.ErrorCode = string(errorResponse.Error.Code)
+
+			if errorResponse.Error.Details != nil {
+				errMap.Details = fmt.Sprintf("%+v", errorResponse.Error.Details)
+			}
+		}
+	}
+
+	errMap.Body = string(resBodyBytes)
+
+	if errMap.Message != "" {
+		errMap.Message = errMsgPrefix + errMap.Message
+	}
+
+	return formatError(errMap)
+}
+
+func formatError(errMap errorResponseMap) error {
```

Comment:
This is somewhat duplicative and I'm open to changing it. I couldn't really land on a good way of trying to represent things.

This basically just turns the map into a JSON string and stuffs it into the `Msg` in `PineconeError`.